File: scripts/build-verlegegebiet-karte.mjs
#!/usr/bin/env node
/**
 * Basiskarten fuer die Sektion "TP Verlegegebiet" rendern.
 *
 * Warum selbst rendern und nicht Kacheln zusammensetzen: fertige Rasterkacheln
 * gibt es frei nur in einfacher Aufloesung, und ihre Schriftgroesse haengt fest
 * an der Zoomstufe. Auf einem Retina-Display sind die Ortsnamen damit entweder
 * unscharf oder zu klein - beides war in der ersten Fassung sichtbar. Hier
 * werden Vektorkacheln lokal in echter doppelter Aufloesung gerendert, und die
 * Schriftgroesse ist eine eigene Entscheidung.
 *
 * Im Shop landet trotzdem nur ein fertiges WebP: zur Laufzeit gibt es keinen
 * externen Request, kein JavaScript fuer die Darstellung, keinen Layout Shift
 * und nichts, was auf dem Smartphone das Scrollen abfangen koennte.
 *
 * Aufruf (braucht Netz und das devDependency puppeteer):
 *     node scripts/build-verlegegebiet-karte.mjs
 *
 * Quellen und Lizenzen:
 *   Kacheln  OpenFreeMap (openfreemap.org) auf Basis von OpenStreetMap, ODbL.
 *            Die Namensnennung steht sichtbar unter der Karte in der Sektion.
 *   Stil     "positron" von OpenFreeMap, hier umgefaerbt auf die Shop-Palette.
 *
 * Aendert sich der Standort oder eine Groesse, muss die Geometrie im Kopf von
 * sections/tp-verlegegebiet.liquid mitgezogen werden - der Radiuskreis rechnet
 * damit und darf nicht raten. Dieses Skript gibt die Werte am Ende aus.
 */
import { createServer } from 'node:http';
import { writeFile } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import puppeteer from 'puppeteer';

const WURZEL = path.dirname(path.dirname(fileURLToPath(import.meta.url)));
const ASSETS = path.join(WURZEL, 'assets');

const LAT = 52.7378;
const LON = 13.2485;   // Saarlandstrasse, 16515 Oranienburg

/**
 * MapLibre rechnet mit 512er-Kacheln, die uebliche Formel mit 256ern - deshalb
 * ist Zoom 8 hier so gross wie Zoom 9 dort. Die Breite ist so gewaehlt, dass
 * der 50-km-Kreis die Flaeche fuellt, ohne anzustossen.
 */
const VARIANTEN = [
  { name: 'desktop', breite: 1000, hoehe: 700, zoom: 8, stadt: 13, ort: 11.5 },
  { name: 'mobile', breite: 340, hoehe: 380, zoom: 7, stadt: 14, ort: 12 },
];

const meterProPixel = (zoom) => (156543.03392 * Math.cos((LAT * Math.PI) / 180)) / 2 ** (zoom + 1);

const SEITE = `<!doctype html>
<meta charset="utf-8">
<title>Kartenrenderer</title>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/maplibre-gl/4.7.1/maplibre-gl.min.css">
<style>html,body{margin:0}#karte{width:100vw;height:100vh}</style>
<div id="karte"></div>
<script src="https://cdnjs.cloudflare.com/ajax/libs/maplibre-gl/4.7.1/maplibre-gl.min.js"></script>
<script>
  const P = new URLSearchParams(location.search);

  // Warme, sehr helle Palette - dieselbe Familie wie im Shop.
  const GRUND = '#faf8f5';
  const WASSER = '#e9e6df';
  const WALD = '#f1efe9';
  const STRASSE = '#eae5dd';
  const SCHRIFT = '#4a443c';

  const karte = new maplibregl.Map({
    container: 'karte',
    style: 'https://tiles.openfreemap.org/styles/positron',
    center: [${LON}, ${LAT}],
    zoom: parseFloat(P.get('zoom')),
    attributionControl: false,
    interactive: false,
    fadeDuration: 0,
  });

  karte.on('load', () => {
    const stil = karte.getStyle();

    // Von der Beschriftung bleiben nur Orte stehen. Strassennummern,
    // Gewaesser- und Flughafennamen beantworten die Frage "wie weit fahren
    // wir" nicht und machen die Flaeche unruhig.
    const ORTSEBENEN = new Set(['label_city', 'label_city_capital', 'label_town']);

    for (const ebene of stil.layers) {
      if (ebene.type !== 'symbol') continue;
      if (!ORTSEBENEN.has(ebene.id)) {
        karte.setLayoutProperty(ebene.id, 'visibility', 'none');
        continue;
      }
      const gross = ebene.id !== 'label_town';
      karte.setLayoutProperty(ebene.id, 'text-size', parseFloat(P.get(gross ? 'stadt' : 'ort')));
      karte.setLayoutProperty(ebene.id, 'text-letter-spacing', 0.01);
      // Oranienburg traegt die Sektion als eigenen Marker - zweimal derselbe
      // Ortsname an derselben Stelle liest sich wie ein Fehler.
      karte.setFilter(ebene.id, ['all', ebene.filter || true, ['!=', ['get', 'name'], 'Oranienburg']]);
      karte.setPaintProperty(ebene.id, 'text-color', SCHRIFT);
      karte.setPaintProperty(ebene.id, 'text-halo-color', GRUND);
      karte.setPaintProperty(ebene.id, 'text-halo-width', 1.6);
    }

    karte.setPaintProperty('background', 'background-color', GRUND);
    for (const ebene of stil.layers) {
      if (ebene.source === 'ne2_shaded' || /shaded|hillshade/.test(ebene.id) || /boundary/.test(ebene.id)) {
        karte.setLayoutProperty(ebene.id, 'visibility', 'none');
      } else if (ebene.type === 'fill' && ebene['source-layer'] === 'water') {
        karte.setPaintProperty(ebene.id, 'fill-color', WASSER);
      } else if (ebene.id.startsWith('landcover') || /park|wood|forest/.test(ebene.id)) {
        if (ebene.type === 'fill') karte.setPaintProperty(ebene.id, 'fill-color', WALD);
      } else if (ebene.type === 'line' && /highway|road|bridge|tunnel/.test(ebene.id)) {
        karte.setPaintProperty(ebene.id, 'line-color', STRASSE);
      } else if (ebene.id === 'waterway') {
        karte.setPaintProperty(ebene.id, 'line-color', WASSER);
      }
    }

    karte.once('idle', () => { window.fertig = true; });
  });
</script>`;

const server = createServer((_, antwort) => {
  antwort.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' });
  antwort.end(SEITE);
});
await new Promise((fertig) => server.listen(0, '127.0.0.1', fertig));
const basis = `http://127.0.0.1:${server.address().port}/`;

// SwiftShader, weil Headless-Chrome hier keinen GPU-Kontext bekommt. Ohne die
// Flags scheitert MapLibre still an "Failed to initialize WebGL".
const browser = await puppeteer.launch({
  headless: 'new',
  args: ['--use-gl=angle', '--use-angle=swiftshader', '--enable-unsafe-swiftshader', '--ignore-gpu-blocklist'],
});

const geometrie = {};
for (const v of VARIANTEN) {
  for (const dichte of [1, 2]) {
    const seite = await browser.newPage();
    const fehler = [];
    seite.on('pageerror', (e) => fehler.push(e.message));
    await seite.setViewport({ width: v.breite, height: v.hoehe, deviceScaleFactor: dichte });
    await seite.goto(`${basis}?zoom=${v.zoom}&stadt=${v.stadt}&ort=${v.ort}`, { waitUntil: 'networkidle0', timeout: 90000 });
    await seite.waitForFunction('window.fertig === true', { timeout: 90000 });
    // Nach "idle" fehlen manchmal noch Schriftglyphen; ohne die Pause bleiben
    // einzelne Ortsnamen leer.
    await new Promise((r) => setTimeout(r, 1500));
    const bild = await seite.screenshot({ type: 'webp', quality: 88 });
    if (fehler.length) throw new Error(`${v.name} @${dichte}x: ${fehler.join(' | ')}`);
    const datei = path.join(ASSETS, `tp-verlegegebiet-${v.name}${dichte === 2 ? '-2x' : ''}.webp`);
    await writeFile(datei, bild);
    console.log(`${path.basename(datei)}  ${v.breite * dichte}x${v.hoehe * dichte}  ${Math.round(bild.length / 1024)} KB`);
    await seite.close();
  }
  const mpp = meterProPixel(v.zoom);
  geometrie[v.name] = { breite: v.breite, hoehe: v.hoehe, meter_pro_pixel: Number(mpp.toFixed(2)) };
  console.log(`   ${v.name}: ${mpp.toFixed(2)} m je Punkt, 50 km = ${((50000 / mpp / v.breite) * 100).toFixed(3)} % der Breite`);
}

await browser.close();
server.close();

console.log('\nGeometrie fuer sections/tp-verlegegebiet.liquid:');
console.log(JSON.stringify(geometrie, null, 2));

File: scripts/build-verlegegebiet-karte.py
#!/usr/bin/env python3
"""Basiskarten und PLZ-Tabelle fuer die Sektion "TP Verlegegebiet" erzeugen.

Warum vorgerendert und nicht als Live-Karte: die Sektion braucht weder Zoom
noch Pan, und ein statisches Bild kostet keinen externen Request (DSGVO),
kein JavaScript, keine Ladezeit und keinen Layout Shift. Die Kacheln werden
einmalig hier geholt, entsaettigt und aufgehellt; im Shop liegt nur das
fertige WebP.

Aufruf (braucht Pillow, nicht Teil der Laufzeit des Shops):
    python3 -m pip install --user Pillow
    python3 scripts/build-verlegegebiet-karte.py

Quellen und Lizenzen:
  Kacheln  OpenStreetMap-Mitwirkende, ODbL - https://osm.org/copyright
           Namensnennung steht sichtbar unter der Karte in der Sektion.
  PLZ      GeoNames postal codes, CC BY 4.0 - https://download.geonames.org
           Namensnennung steht im Feld "quelle" der erzeugten JSON-Datei.

Aendert sich der Standort oder der maximale Radius, hier die Konstanten
anpassen und neu erzeugen - die Sektion rechnet den Kreis aus GEOMETRIE
und darf nicht raten.
"""
import io
import json
import math
import os
import unicodedata
import urllib.request
import zipfile
from collections import defaultdict

from PIL import Image, ImageFilter, ImageOps

# --- Standort -------------------------------------------------------------
LAT, LON = 52.7378, 13.2485          # Saarlandstrasse, 16515 Oranienburg
UA = "TeppichParadies-Theme/1.0 (kontakt@teppich-paradies.net)"

# --- Kartengeometrie ------------------------------------------------------
# Zoomstufe und Bildgroesse in logischen Kartenpixeln. Die Sektion braucht
# daraus nur das Seitenverhaeltnis und die Meter pro Pixel; beides wird unten
# ausgegeben und steht als Kommentar in sections/tp-verlegegebiet.liquid.
# Der Ausschnitt ist bewusst knapp: der Kreis soll die Flaeche fuellen, ohne
# anzustossen. Breiter gewaehlt zeigt die Karte halb Brandenburg und der Kreis
# verliert seine Aussage. Auf dem Smartphone kommt dazu, dass die Kacheln ihre
# Schriftgroesse fest mitbringen: je knapper der Ausschnitt, desto groesser
# erscheinen die Ortsnamen auf dem schmalen Display. Die mobile Breite liegt
# nahe an der tatsaechlichen Anzeigebreite eines Telefons, damit das Bild dort
# kaum hochskaliert werden muss.
VARIANTEN = (
    # name       zoom  breite  hoehe  schwelle  schrift
    ("desktop",     9,   1000,    700,     112,    0.64),
    ("mobile",      8,    340,    380,      80,    0.55),
)

# --- Umfaerbung -----------------------------------------------------------
# Getrennte Behandlung ist der Kern: Flaechen (Wald, Wasser, Siedlung) werden
# fast auf Weiss gezogen, die dunklen Ortsnamen bleiben lesbar. Ohne die
# Trennung ist die Karte entweder ruhig und unlesbar oder lesbar und unruhig.
# Die Schwelle steht je Variante in VARIANTEN. Auf dem Smartphone ist sie
# bewusst niedriger: dort sind die Namen kleiner Orte nur noch sieben Pixel
# hoch und zerfallen zu grauem Rauschen. Eine niedrige Schwelle laesst genau
# diese Kantenglaettung ins Weisse laufen und behaelt nur die Kerne der
# grossen Namen - Berlin und Potsdam bleiben scharf, der Rest verschwindet.
# Wenige lesbare Namen sind mehr wert als zwanzig unlesbare.
FLAT = 0.16          # Restanteil der Flaechenzeichnung
HELL = (250, 248, 245)   # Grundton hell: warmes Off-White wie im Shop
DUNKEL = (28, 26, 24)    # Grundton dunkel: warmes Anthrazit

# Die Kacheln gibt es nur in einfacher Aufloesung, und die Schriftgroesse haengt
# an der Zoomstufe: eine hoehere Stufe waere schaerfer, ihre Ortsnamen waeren auf
# der fertigen Karte aber halb so gross und damit unlesbar. Deshalb bleibt die
# Stufe, und die 2x-Fassung wird hier hochgerechnet und nachgeschaerft - das ist
# sichtbar sauberer als die weiche Interpolation des Browsers auf Retina.
SCHAERFE = dict(radius=1.1, percent=110, threshold=2)

# --- PLZ-Tabelle ----------------------------------------------------------
GEONAMES = "https://download.geonames.org/export/zip/DE.zip"
MAX_KM = 90          # Reserve ueber dem im Editor einstellbaren Radius (max. 60)

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
ASSETS = os.path.join(ROOT, "assets")
CACHE = os.path.join(ROOT, ".cache", "osm-tiles")


def lonlat_to_px(lon, lat, z, tile_px=256):
    """Weltpixel-Koordinaten in Web Mercator."""
    n = 2 ** z * tile_px
    s = math.sin(math.radians(lat))
    return ((lon + 180.0) / 360.0 * n,
            (0.5 - math.log((1 + s) / (1 - s)) / (4 * math.pi)) * n)


def meter_pro_pixel(z):
    return 156543.03392 * math.cos(math.radians(LAT)) / 2 ** z


def kachel(z, x, y):
    os.makedirs(CACHE, exist_ok=True)
    pfad = os.path.join(CACHE, f"osm-{z}-{x}-{y}.png")
    if not os.path.exists(pfad):
        req = urllib.request.Request(f"https://tile.openstreetmap.org/{z}/{x}/{y}.png",
                                     headers={"User-Agent": UA})
        with open(pfad, "wb") as f:
            f.write(urllib.request.urlopen(req, timeout=30).read())
    return Image.open(pfad).convert("RGB")


def stitchen(z, breite, hoehe, tp=256):
    cx, cy = lonlat_to_px(LON, LAT, z, tp)
    links, oben = cx - breite / 2, cy - hoehe / 2
    x0, y0 = int(links // tp), int(oben // tp)
    x1, y1 = int((links + breite) // tp), int((oben + hoehe) // tp)
    flaeche = Image.new("RGB", ((x1 - x0 + 1) * tp, (y1 - y0 + 1) * tp))
    for x in range(x0, x1 + 1):
        for y in range(y0, y1 + 1):
            flaeche.paste(kachel(z, x, y), ((x - x0) * tp, (y - y0) * tp))
    ox, oy = int(links - x0 * tp), int(oben - y0 * tp)
    return flaeche.crop((ox, oy, ox + breite, oy + hoehe))


def beruhigen(img, schwelle, schrift):
    def ton(x):
        return x * schrift if x < schwelle else 255 - (255 - x) * FLAT
    grau = ImageOps.grayscale(img).point(
        lambda x: max(0, min(255, int(round(ton(x)))))
    )
    return ImageOps.colorize(grau, DUNKEL, HELL)


def karten_bauen():
    geometrie = {}
    for name, z, breite, hoehe, schwelle, schrift in VARIANTEN:
        karte = beruhigen(stitchen(z, breite, hoehe), schwelle, schrift)
        ziel = os.path.join(ASSETS, f"tp-verlegegebiet-{name}.webp")
        karte.save(ziel, "WEBP", quality=85, method=6)
        ziel2x = os.path.join(ASSETS, f"tp-verlegegebiet-{name}-2x.webp")
        (karte.resize((breite * 2, hoehe * 2), Image.LANCZOS)
              .filter(ImageFilter.UnsharpMask(**SCHAERFE))
              .save(ziel2x, "WEBP", quality=82, method=6))
        mpp = meter_pro_pixel(z)
        geometrie[name] = {
            "breite": breite, "hoehe": hoehe, "meter_pro_pixel": round(mpp, 2),
            "km_je_prozent_breite": round(breite * mpp / 100000, 4),
        }
        print(f"{os.path.basename(ziel)}  {breite}x{hoehe}  "
              f"{os.path.getsize(ziel) / 1024:.0f} KB  (2x: "
              f"{os.path.getsize(ziel2x) / 1024:.0f} KB)  {mpp:.2f} m/px  "
              f"50 km = {50000 / mpp / breite * 100:.3f} % der Breite")
    return geometrie


def haversine(lat1, lon1, lat2, lon2):
    r = 6371.0088
    p1, p2 = math.radians(lat1), math.radians(lat2)
    a = (math.sin((p2 - p1) / 2) ** 2
         + math.cos(p1) * math.cos(p2) * math.sin(math.radians(lon2 - lon1) / 2) ** 2)
    return 2 * r * math.asin(math.sqrt(a))


def normalisieren(s):
    s = s.lower().replace("ß", "ss").replace("ä", "ae").replace("ö", "oe").replace("ü", "ue")
    s = unicodedata.normalize("NFKD", s)
    return "".join(c for c in s if c.isalnum())


def plz_tabelle_bauen():
    daten = urllib.request.urlopen(
        urllib.request.Request(GEONAMES, headers={"User-Agent": UA}), timeout=60).read()
    with zipfile.ZipFile(io.BytesIO(daten)) as z:
        zeilen = z.read("DE.txt").decode("utf-8").splitlines()

    plz = {}
    # Grossstaedte reichen ueber den Rand, deshalb je Ort die naechste und die
    # entfernteste PLZ - sonst meldet "Berlin" pauschal "im Gebiet".
    orte = defaultdict(lambda: None)
    for zeile in zeilen:
        c = zeile.split("\t")
        if len(c) < 11 or not c[9] or not c[10]:
            continue
        try:
            km = haversine(LAT, LON, float(c[9]), float(c[10]))
        except ValueError:
            continue
        if km > MAX_KM:
            continue
        code, ort = c[1], c[2]
        if code not in plz or km < plz[code]:
            plz[code] = km
        n = normalisieren(ort)
        if not n:
            continue
        spanne = orte[n]
        orte[n] = [km, km] if spanne is None else [min(spanne[0], km), max(spanne[1], km)]

    ziel = os.path.join(ASSETS, "tp-verlegegebiet-orte.json")
    with open(ziel, "w", encoding="utf-8") as f:
        json.dump({
            "quelle": "GeoNames postal codes, CC BY 4.0, https://www.geonames.org",
            "mittelpunkt": {"ort": "Oranienburg", "lat": LAT, "lon": LON},
            "max_km": MAX_KM,
            "plz": {k: round(v, 1) for k, v in sorted(plz.items())},
            "orte": {k: [round(v[0], 1), round(v[1], 1)] for k, v in sorted(orte.items())},
        }, f, ensure_ascii=False, separators=(",", ":"))
    print(f"{os.path.basename(ziel)}  {len(plz)} PLZ, {len(orte)} Orte, "
          f"{os.path.getsize(ziel) / 1024:.0f} KB")


if __name__ == "__main__":
    geometrie = karten_bauen()
    plz_tabelle_bauen()
    print("\nGeometrie fuer sections/tp-verlegegebiet.liquid:")
    print(json.dumps(geometrie, indent=2, ensure_ascii=False))
